cozy/structures/heaps.py

```diff
@@ -168,7 +196,7 @@ def storage_size(self, e : Exp, storage_size) -> Exp:
         storage sizes for other expressions.
         """
         assert type(e.type) in (TMinHeap, TMaxHeap)
-        return storage_size(EHeapElems(e).with_type(TBag(e.type.elem_type)))
+        return storage_size(shallow_copy(heap_elems(e)).with_type(TBag(e.type.elem_type)))
```

Comment:
This is extremely fishy. You should always do the copy where you modify the object, rather than making the contract of the `heap_elems` method that callers need to make a copy before calling it.